crates/pink/src/local_cache.rs

```diff
@@ -0,0 +1,220 @@
+
+#[derive(Default, Debug)]
+struct Storage {
+    // Sum of the size of all the keys and values.
+    size: usize,
+    kvs: HashMap<Vec<u8>, StorageValue>,
+}
+
+#[derive(Debug)]
+struct StorageValue {
+    // Expiration time in seconds since the first call to `now`.
+    expire_at: u64,
+    value: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct LocalCache {
+    // Number of set ops between two GC ops.
+    gc_interval: u64,
+    // Accumulated number of set ops since last GC.
+    sets_since_last_gc: u64,
+    // Default expiration time in seconds.
+    default_value_lifetime: u64,
+    max_cache_size_per_contract: usize,
+    storages: HashMap<Vec<u8>, Storage>,
+}
+
+impl Default for LocalCache {
+    fn default() -> Self {
+        Self {
+            gc_interval: 1000,
+            sets_since_last_gc: 0,
+            default_value_lifetime: 3600 * 24 * 7, // 1 week
+            max_cache_size_per_contract: 10 * 1024 * 1024, // 10MB
+            storages: Default::default(),
+        }
+    }
+}
+
+impl LocalCache {
+    fn maybe_clear_expired(&mut self) {
+        self.sets_since_last_gc += 1;
+        if self.sets_since_last_gc == self.gc_interval {
+            self.sets_since_last_gc = 0;
+            let now = now();
+            self.storages.values_mut().for_each(|storage| {
+                let storage_size = &mut storage.size;
+                storage.kvs.retain(|k, v| {
+                    if v.expire_at > now {
+                        true
+                    } else {
+                        *storage_size -= v.value.len() + k.len();
+                        false
+                    }
+                });
+            });
+        }
+    }
+
+    pub fn get(&self, id: &[u8], key: &[u8]) -> Option<Vec<u8>> {
+        Some(self.storages.get(id)?.kvs.get(key)?.value.to_owned())
+    }
+
+    pub fn set(
+        &mut self,
+        id: Cow<[u8]>,
+        key: Cow<[u8]>,
+        value: Cow<[u8]>,
+    ) -> Result<(), StorageQuotaExceeded> {
+        self.maybe_clear_expired();
+        let store = self
+            .storages
+            .entry(id.into_owned())
+            .or_insert_with(Storage::default);
+        let key_len = key.len();
+        let value_len = value.len();
+        let prev_value = store.kvs.remove(key.as_ref());
+
+        let new_size = match prev_value {
+            Some(v) => store.size + value_len - v.value.len(),
+            None => store.size + key_len + value_len,
+        };
+
+        if new_size > self.max_cache_size_per_contract {
+            return Err(StorageQuotaExceeded);
+        }
+
+        store.size = new_size;
+        store.kvs.insert(
+            key.into_owned(),
+            StorageValue {
+                expire_at: now().saturating_add(self.default_value_lifetime),
+                value: value.into_owned(),
+            },
+        );
+        Ok(())
+    }
+
+    pub fn set_expire(&mut self, id: Cow<[u8]>, key: Cow<[u8]>, expire: u64) {
+        self.storages
```

Comment:
What if `expire` is earlier than the current time? Should we just drop it?

Reply:
It might make not much difference. GCs in programing languages without refcount delays the destruction to next recycling turn. But it's good to drop so we can free the memory earlier.